overlord/snapstate/handlers_components.go

```diff
@@ -615,35 +615,49 @@ func (m *SnapManager) doRemoveKernelModulesSetup(t *state.Task, finalStatus stat
 	return nil
 }
 
-func infoForCompUndo(t *state.Task) (*snap.ComponentSideInfo, string, error) {
+func infoForCompUndo(t *state.Task) (*snap.ComponentSideInfo, *snap.ComponentSideInfo, string, snap.Revision, error) {
 	st := t.State()
 	st.Lock()
 	defer st.Unlock()
 
-	_, snapsup, err := TaskComponentSetup(t)
+	compsup, snapsup, err := TaskComponentSetup(t)
 	if err != nil {
-		return nil, "", err
+		return nil, nil, "", snap.R(0), err
 	}
 
 	setupTask, err := componentSetupTask(t)
 	if err != nil {
-		return nil, "", err
+		return nil, nil, "", snap.R(0), err
 	}
 	var unlinkedComp sequence.ComponentState
 	err = setupTask.Get("unlinked-component", &unlinkedComp)
 	if err != nil {
-		return nil, "", fmt.Errorf("internal error: no component to discard: %w", err)
+		return nil, nil, "", snap.R(0),
+			fmt.Errorf("internal error: no component to discard: %w", err)
 	}
 
-	return unlinkedComp.SideInfo, snapsup.InstanceName(), nil
+	return unlinkedComp.SideInfo, compsup.CompSideInfo,
+		snapsup.InstanceName(), snapsup.Revision(), nil
 }
 
 func (m *SnapManager) doDiscardComponent(t *state.Task, _ *tomb.Tomb) error {
-	compSideInfo, instanceName, err := infoForCompUndo(t)
+	unlinkedSi, snapsupSi, instanceName, snapRev, err := infoForCompUndo(t)
 	if err != nil {
 		return err
 	}
 
+	// If revisions for compsetup/unlinked component are the same then this
+	// is not replacing a component, but removing one (would happen on snap
+	// removal). In that case we want to remove the symlink to the active
+	// component revision too.
+	if snapsupSi.Revision == unlinkedSi.Revision {
+		cpi := snap.MinimalComponentContainerPlaceInfo(unlinkedSi.Component.ComponentName,
+			unlinkedSi.Revision, instanceName)
+		if err := m.backend.UnlinkComponent(cpi, snapRev); err != nil {
```

Comment:
this is a bit strange, don't we have a separate task about this ?

Reply:
does discard-snap have similar behavior?

Reply:
We did not have a task that could handle unlinking a component for a snap revision, only for the current snap, that is why I was hesitant on creating more tasks for this. But in the end I went ahead and created a new `unlink-component` task for this so thing are more coherent.